### ISSUE TITLE
1438 TLE with Kotlin

### ISSUE THREAD

**Issue by Tony080** [LeetCode-Feedback/LeetCode-Feedback#2594]
#### Your LeetCode username

AzraelZealous

#### Category of the bug
- [ ] Question
- [ ] Solution
- [x] Language

#### Description of the bug
Same code can be submitted with Java but got TLE with Kotlin

##### Java code
```
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        Deque<Integer> maxs = new ArrayDeque<>();
        Deque<Integer> mins = new ArrayDeque<>();
        
        int left = 0;
        int right = 0;
        while (right < nums.length) {
            while (!maxs.isEmpty() && maxs.peekLast() < nums[right]) {
                maxs.pollLast();
            }
            while (!mins.isEmpty() && mins.peekLast() > nums[right]) {
                mins.pollLast();
            }
            maxs.addLast(nums[right]);
            mins.addLast(nums[right]);
            if (maxs.peek() - mins.peek() > limit) {
                if (maxs.peek() == nums[left]) {
                    maxs.removeFirst();
                }
                if (mins.peek() == nums[left]) {
                    mins.removeFirst();
                }
                left++;
            }
            right++;
        }
        return right - left;
    }
}
```

##### Kotlin code **Got TLE**
```
class Solution {
    fun longestSubarray(nums: IntArray, limit: Int): Int {
        val maxs = ArrayDeque<Int>()
        val mins = ArrayDeque<Int>()
        
        var left = 0
        var right = 0
        while (right < nums.size) {
            while (maxs.isNotEmpty() && maxs.last() < nums[right]) {
                maxs.removeLast()
            }
            while (mins.isNotEmpty() && mins.last() > nums[right]) {
                mins.removeLast()
            }
            maxs.addLast(nums[right])
            mins.addLast(nums[right])
            if (maxs.first() - mins.first() > limit) {
                if (maxs.first() == nums[left]) {
                    maxs.removeFirst()
                }
                if (mins.first() == nums[left]) {
                    mins.removeFirst()
                }
                left++
            }
            right++
        }
        return right - left
    }
}
```

#### Language used for code
Kotlin

#### Expected behavior
This is the algorithm with O(n) time complexity which is the optimal answer, should pass all the cases.

#### Screenshots

N/A

#### Additional context

N/A

**Comment by lc-kay** [comment]
Hi @Tony080 
Thank you for reaching out to us. I've relayed this issue to our team to investigate.

**Comment by lc-kay** [comment]
@Tony080 Our team has reviewed your report and concluded that the provided solution is not O(n) and performs slow. Please refer to the documentation of the Kotlin language to read more about the deque and implement a true O(n) solution.

**Comment by Tony080** [comment]
@lc-kay Why you said it's not O(n)? Both Java ArrayDeque and Kotlin ArrayDeque are implemented by using circular array with O(1) time complexity for enque/deque on both end. 

Is the any reference to make the conclusion it's not an O(n) solution?

**Comment by lc-kay** [comment]
@Tony080 
<img width="1206" alt="Screen Shot 2021-03-10 at 12 26 19 PM" src="https://user-images.githubusercontent.com/70870126/110693037-dc3f9300-819b-11eb-83d1-b5103a22d9f5.png">
<img width="1188" alt="Screen Shot 2021-03-10 at 12 26 12 PM" src="https://user-images.githubusercontent.com/70870126/110693040-dd70c000-819b-11eb-95b1-3937455a022c.png">

Here's an O(n) Kotlin code that can AC in 500ms and the time limit is 5000ms. Please check out this solution for reference.

**Comment by Tony080** [comment]
@lc-kay 
Thanks for the screenshot. 

I found the root cause. LeetCode is importing `java.util.ArrayDeque`(perhaps `java.util.*`). And, the function `last()` and `first()` are not built natively in Java. So, [The Kotlin Standard Library](https://github.com/JetBrains/kotlin/tree/master/libraries/stdlib) is used to provide those as extension functions.

As I tested on my machine with one million 1's as input, there are three different results shown
```
Java ArrayDeque with GenerateStandardLib 106
Pure Kotlin ArrayDeque 26
Pure Java ArrayDeque 26
```

The test code has been uploaded [here](https://github.com/Tony080/Kotlin-1438-demo/blob/main/main.kt).
So the native Kotlin ArrayDeque has similar performance as Java has. Just the redundant java utilities import statements make things worse. 

In my opinion, LeetCode should remove those java utility imports, and encourage people who submit code in Kotlin to use native Kotlin data structures instead.
